File: src/1.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> retVal;

        unordered_map<int, int> hashTable;
        int numsSize = nums.size();
        for (int i = 0; i < numsSize; ++i) {
            auto iterator = hashTable.find(target - nums[i]);
            if (iterator != hashTable.end()) {
                retVal.push_back(iterator->second);
                retVal.push_back(i);
                return retVal;
            }
            hashTable[nums[i]] = i;
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<int> nums;
        int target;
    };
    vector<subject> testData{{{2, 7, 11, 15}, 9}, {{3, 2, 4}, 6}, {{3, 3}, 6}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: nums = [2,7,11,15], target = 9
     *  Output: [0,1]
     *
     *  Input: nums = [3,2,4], target = 6
     *  Output: [1,2]
     *
     *  Input: nums = [3,3], target = 6
     *  Output: [0,1]
     */

    Solution cSolution;
    vector<int> answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: nums = [";
        for (long unsigned int j = 0; j < testData[i].nums.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << testData[i].nums[j];
        }
        cout << "], target = " << testData[i].target << "\n";

        answer = cSolution.twoSum(testData[i].nums, testData[i].target);
        cout << "Output: [";
        for (long unsigned int j = 0; j < answer.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << answer[j];
        }
        cout << "]\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/1.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        retVal = []

        hashTable = defaultdict(int)
        for idx, num in enumerate(nums):
            if num in hashTable:
                retVal = [hashTable[num], idx]
                break
            hashTable[target - num] = idx

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums, target in [[[2, 7, 11, 15], 9], [[3, 2, 4], 6], [[3, 3], 6]]:
            # /* Example
            #  *  Input: nums = [2,7,11,15], target = 9
            #  *  Output: [0,1]
            #  *
            #  *  Input: nums = [3,2,4], target = 6
            #  *  Output: [1,2]
            #  *
            #  *  Input: nums = [3,3], target = 6
            #  *  Output: [0,1]
            #  */
            logging.info("Input: nums = %s, target = %s", nums, target)

            retVal = pSolution.twoSum(nums, target)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
